webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js

```diff
@@ -93,7 +93,7 @@ describe('medic-xpath-extensions', function () {
       ['2015-10-01T11:11:11.111', '2014-10-01T11:11:11.111', -12,],
       ['2014-10-01T00:00:00.000', '2015-10-01T00:00:00.000', 12,],
       ['August 19, 1975 00:00:00 GMT', 'August 18, 1976 23:15:30 GMT+07:00', 11],
-      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2300', 0],
+      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 21:59:59 GMT+2300', 0],
```

Comment:
For reasons that my weak brain cannot comprehend, in the `Pacific/Auckland` TZ, this would fail, saying the dates are `1` month apart (instead of `0`). Backing it up by a second makes this test pass in all TZ's.  Like I mentioned in the description, I think this is acceptable since all of this is just based on Moment's logic...

Reply:
:sweat:  I think I have finally determined the cause of this behavior.  Basically I think it comes down to Moment's magical handling of [Month and year diffs](https://momentjs.com/docs/#/displaying/difference/)

> moment#diff has some special handling for month and year diffs. It is optimized to ensure that two months with the same date are always a whole number apart.
>
>So Jan 15 to Feb 15 should be exactly 1 month.
>
>Feb 28 to Mar 28 should be exactly 1 month.
>
>Feb 28 2011 to Feb 28 2012 should be exactly 1 year.

On the surface this seem fine and you would not think that you would get inconsistent behavior here _when your date values are declared with explicit time zones_.  However, I think that Moment is going to evaluate this magical date logic based on day/month values for the dates _in the timezone of the runtime environment._  

In `GMT`, these dates resolve to `Sun, 25 Sep 2005 00:00:00 GMT` and `Mon, 24 Oct 2005 23:00:00 GMT` which should clock in as an hour less than `1 month` (if both resolved to the `25th` day of the month, then the Moment magic would kick in and say they are `1 month` apart).  But, when running in an environment where the TZ is set to  `Pacific/Auckland` we need to roll the dates `+12` which gets us: `Sun, 25 Sep 2005 12:00:00 GMT+1200` `Tues, 25 Oct 2005 11:00:00 GMT+1200`.  Now they are on the same day, but the time of the second date is still _before the first_.  So, how are they a month apart?  Yep, you guessed it.  In 2005, DST for `Pacific/Auckland` started on 2 Oct!  That gives the extra hour which bumps this to exactly 1 month apart (according to the Moment magic).... :facepalm:

Reply:
Ahh DST. That old chestnut...